src/hotspot/share/opto/vectorization.hpp

```diff
@@ -259,4 +263,322 @@ class VectorElementSizeStats {
   }
 };
 
+// When alignment is required, we must adjust the pre-loop iteration count pre_iter,
+// such that the address is aligned for any main_iter >= 0:
+//
+//   adr = base + offset + invar + scale * init
+//                               + scale * pre_stride * pre_iter
+//                               + scale * main_stride * main_iter
+//
+// The AlignmentSolver generates solutions of the following forms:
+//   1. Empty with a failure reason.
+//   2. Trivial (any pre-loop limit guarantees alignment).
+//   3. Constrained (r, q, mem_ref, scale, invar)
+//        Where scale is 0 if no scale dependency,
+//        and invar is nullptr if no invar dependency.
+//
+// The Constrained solution is of the following form:
+//
+//   pre_iter = pre_iter_C_const  + pre_iter_C_init              + pre_iter_C_invar
+//            = pre_r + pre_q * m + alignment_init(X * var_init) + alignment_invar(Y * var_invar)
+//
+// Essentially, we get a periodic solution for the C_const term, which
+// is shifted by the init and the invar terms. For two solutions to be
+// compatible, they must have the same periodic part (or subset), which
+// we can easily check with p an q. Compatible solutions must also have
+// the same shifting, which we can check with the scale and invar dependency.
+//
+class AlignmentSolutionEmpty;
+class AlignmentSolutionTrivial;
+class AlignmentSolutionConstrained;
```

Comment:
I usually prefer to have the other way round to better distinguish the different classes in the code. But that's just personal taste. I leave it up to you to decide which one you like better.
```suggestion
class EmptyAlignmentSolution;
class TrivialAlignmentSolution;
class ConstrainedAlignmentSolution;
```